pkg/ingester/ingester.go

```diff
@@ -241,13 +250,47 @@ func (i *Ingester) getOrCreateInstance(instanceID string) *instance {
 
 // Query the ingests for log streams matching a set of matchers.
 func (i *Ingester) Query(req *logproto.QueryRequest, queryServer logproto.Querier_QueryServer) error {
-	instanceID, err := user.ExtractOrgID(queryServer.Context())
+	// initialize stats collection for ingester queries and set grpc trailer with stats.
+	ctx := stats.NewContext(queryServer.Context())
+	defer stats.SendAsTrailer(ctx, queryServer)
+
+	instanceID, err := user.ExtractOrgID(ctx)
 	if err != nil {
 		return err
 	}
 
 	instance := i.getOrCreateInstance(instanceID)
-	return instance.Query(req, queryServer)
+	itr, err := instance.Query(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	defer helpers.LogError("closing iterator", itr.Close)
+
+	if i.cfg.QueryStore {
+		start := req.Start
+		end := req.End
```

Comment:
I suggest a little function to create a new request for the store instead of mutating the one you received.

This will also come handy for writing a test.